Deprecate groupby/pivot observed=False default

jseabold opened the pull request:
- [x] Relates to #17594, Closes #30552
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Had a relatively small 70k data frame that I was trying to do a groupby sum on blow up on me today. This was the reason. I had something like zip codes and cities as categoricals, expected SQL-like groupby but instead got a cartesian product of 'cities' and 'zips'. Sounds like there was some previous desire to explore a new default.

Didn't try to do any wild stuff to keep up with the stacklevel depending on where this was called from.

jseabold committed:
Deprecate default observed=False for groupby

jseabold committed:
Deprecate default for series

jseabold committed:
Test warns for series

jseabold committed:
Test that pivot warns

jseabold committed:
Change default observed for pivot table

jseabold committed:
Don't set stacklevel bc of multiple entrypoints

jseabold committed:
Don't check stacklevel on tests.

This code can be reached from multiple places. Keeping up with the
stacklevel seems like overkill here.

jseabold committed:
Blacken

jseabold committed:
Add deprecation doc

jreback commented:
hmm i thought we had an issue about

jseabold commented:
Will see about avoiding `pytest.warns` though the CI stuff doesn't make it real clear why. I think the included warnings assertions didn't like asserting there wasn't a warning but I'll check again. 

First question is, do you want this PR? Does it need discussion?

Second question is, what do y'all suggest I do about the warnings coming in the examples? Suppress them? Fix the examples to use use the new keyword?

jreback commented:
we don't use pytest.warn instead use tm.assert_produces_warning

i think the intent of the PR is good - haven't looked closely yet.

jseabold committed:
Don't use pytest.warns

TomAugspurger commented:
https://github.com/pandas-dev/pandas/issues/30552 is the related issue.

I'm unsure about how to proceed here. We've overloaded `Categorical` for two purposes: 1: a statistical concept for the fixed set of categories, and 2: the memory-savings for low-cardinality data. I think that `observed=True` is a good default for Categoricals and a bad default for memory-saving. I think my ideal outcome is to

1. implement a dtype that just does the dictionary encoding (https://github.com/pandas-dev/pandas/issues/20899).
2. Change the default `observed=False` to `observed=None` to mean "ask the extension type for its default." Then `Categorical` can implement `observed=False` by default and the Dictionary-encoded type can have `observed=True`.

But that's also much more work that this PR, so like I said, I'm unsure how to proceed.

jseabold commented:
Yeah, that makes sense as a better solution, and I was of two minds about whether to do this, but I struggled to find an example for when I'd ever want `observed=False`. Do you have one in mind? If I really wanted a cartesian product of things I don't actually have in my data, I don't _think_ I'd reach for groupby to do that.

jseabold commented:
I started fixing up the doc warnings, but I think there's some more I need to think through with crosstab and `dropna`.

jseabold commented:
Yeah, this is basically my position.

https://github.com/pandas-dev/pandas/issues/30552#issuecomment-602849215

jseabold committed:
okwarning option

jseabold commented:
Rebased to get rid of the merge conflict. Not sure why the coverage tests are saying I added untested lines.

jseabold commented:
Going to be a bit of (tedious) work to get the tests and doc builds passing on the warnings as errors runs, let me know whether this is likely to get merged, and I'll come back and fix the tests and look at the crosstab stuff.

TomAugspurger commented:
I'd say something like 50-80% probability of being accepted? As you say, it's only sometimes where this behavior is desired for (statistical) categorical columns, and it's never desired for the memory-savings purpose.

cc @jankatins, since I suspect this goes back to the original categorical implementation, if you have thoughts.

jankatins commented:
Yes, the original usecase I had in mind was a survey with lots of likert like scales: "Strong Disagree ... neutral ... Strong Agree". The original categorical was also build around what R does for factors. All the "problems" started o surface when someone discovered that categoricals save memory and time when dealing with strings. 

The basic group by was "aggregator(num_col) per cat_column_y" and it should produce the same structure (ordering,number of rows) no matter if cat_column_y contained all values or not (so NA/0), e.g. to get nice plots which look structural similar in a report. For the same reason I would guess it was decided that group bys with two cat columns should show all combinations.

jseabold commented:
> All the "problems" started o surface when someone discovered that categoricals save memory and time when dealing with strings.

Just want to gently push back on the (maybe perceived) notion that this is somehow an abuse of categoricals. High-cardinality, non-independent categoricals/factors are definitely a thing, and the default can not only explode memory but also gives non-sensical answers.

I think the departure from (the flavors of) SQL(s I use) was more unexpected for me.

Buut, as I mentioned in the issue thread, I've definitely wanted both behaviors of this just today. Is there some design philosophy here to fallback to for guidance (like prefer standard SQL semantics or refuse the temptation to guess or something)? I'm not sure just a type with sensible defaults is going to solve the issue for me. E.g., I want the `observed=True` behavior for two independent factors but not for two dependent factors and only when they'd both be included in a groupby. It's the presence of another one that leads me to want the behavior.

All this said, now that I know and am thinking about these things I think I'll always have to specify observed. Part of me just wants to raise an error if `observed=None` and force folks to think, but that'd be a pretty bad UX.

jankatins commented:
For me this feels a bit like the "stringsAsFactors" saga on R: https://simplystatistics.org/2015/07/24/stringsasfactors-an-unauthorized-biography/ :-)

jseabold commented:
> For me this feels a bit like the "stringsAsFactors" saga on R

Ha, indeed. Now you've got me rethinking everything.

jseabold commented:
Been thinking about this a bit, given the comments and use cases. I have a proposal that may or may not be a good one. 

What about making the default None and in the presence of a categorical, if the default isn't changed it raises an error with a message that makes users choose True or False. 

This would be noisy but would avoid the temptation to guess and whatever "my way is the most typical" bias that could creep in. It wouldn't force a "stringsAsFactors" situaish just an "ugh SettingWithCopyWarning" situaish (which I can live with). It also could be temporary until there's another extension dtype ready that's more appropriate for situations where your strings aren't factors (or your factors aren't independent). 

Thoughts?

TomAugspurger commented:
>  What about making the default None and in the presence of a categorical, if the default isn't changed it raises an error with a message that makes users choose True or False.

That sounds fine to me (with the default `None` being a warning for now, saying that an exception will be raised in the future).

jreback commented:
@jseabold yeah sorry haven't gotten to this, but if you can merge master and implement @TomAugspurger suggestion would be great.

jseabold commented:
Yeah, sounds good.

jreback commented:
@jseabold if you have a chance to merge master and fix this up

jseabold committed:
Remove incorrect series note

jseabold committed:
Hoop jump

jseabold commented:
```
Traceback (most recent call last):
  File "/home/runner/work/pandas/pandas/scripts/check_for_inconsistent_pandas_namespace.py", line 59, in <module>
    main()
  File "/home/runner/work/pandas/pandas/scripts/check_for_inconsistent_pandas_namespace.py", line 49, in main
    raise AssertionError(
AssertionError: Found both `pd.DataFrame` and `DataFrame` in pandas/tests/groupby/test_grouping.py
```

lol. Come on!

I rebased on master. One thing this brought to my attention is that there's not a consistent way to choose the behavior you want for methods that rely on the underlying code. E.g., `Series.value_counts`, `DataFrame.sum`, etc.

Cleaning up the docs, I'm also not sure that `pivot_table` does what I expect it do in master. This is master. Wouldn't you expect there to be a c for A in the below? Regardless this PR retains the behavior in master. It now warns but the observed keyword doesn't appear to actually do anything unless I'm missing something.

```python
In [1]: paste
    raw_cat = pd.Categorical(["a", "a", "b", "b"], categories=["a", "b", "c"])
    df = pd.DataFrame({"A": raw_cat, "B": ["c", "d", "c", "d"], "values": [1, 2, 3, 4]})
    pd.pivot_table(df, values="values", index=["A", "B"])
## -- End pasted text --
Out[1]:
     values
A B
a c       1
  d       2
b c       3
  d       4

In [2]: pd.pivot_table(df, values='values', index=["A", "B"], observed=True)
Out[2]:
     values
A B
a c       1
  d       2
b c       3
  d       4

In [3]: pd.pivot_table(df, values='values', index=["A", "B"], observed=False)
Out[3]:
     values
A B
a c       1
  d       2
b c       3
  d       4
```

jseabold committed:
Blacken

jseabold committed:
Fix backticks

jseabold commented:
> It now warns but the observed keyword doesn't appear to actually do anything unless I'm missing something.

Oh, you're supposed to use `dropna` for this. I guess the warning isn't general enough given that the keywords that control this behavior are different depending on entrypoint.

jseabold committed:
okwarning

jseabold committed:
Move method to right class. Rebase mistake.

jseabold commented:
> Oh, you're supposed to use dropna for this and the keywords interact. I guess the warning isn't general enough given that the keywords that control this behavior are different depending on entrypoint. The observed keyword here only matters if dropna is False.
>
> Is there ever a case where dropna is True and observed should be False, or can we just always set observed to False and avoid the warning in the default case?

Answering my own question. Somewhere along the line observed=False stopped returning NaNs and started returning 0s. Didn't see when this happened glancing at the release notes between [v0.23.0](https://pandas.pydata.org/docs/whatsnew/v0.23.0.html#new-observed-keyword-for-excluding-unobserved-categories-in-groupby) and master.

jseabold committed:
Expect FutureWarning

pep8speaks commented:
Hello @jseabold! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:













* In the file [`pandas/tests/groupby/aggregate/test_cython.py`](https://github.com/pandas-dev/pandas/blob/bae44c09549ed010059e6059862eeda06afb92de/pandas/tests/groupby/aggregate/test_cython.py):

> [Line 186:89](https://github.com/pandas-dev/pandas/blob/bae44c09549ed010059e6059862eeda06afb92de/pandas/tests/groupby/aggregate/test_cython.py#L186): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)

jseabold committed:
crosstab doesn't have observed keyword

jseabold commented:
Finished the warnings whack-a-mole. Silencing warnings in years old what's new docs seemed like not the right thing to do, but it's in there. I don't know if you re-publish these and/or accept ipython directives in them, but if so they're likely not historically correct anymore.

jreback commented on doc/source/user_guide/10min.rst:
```diff
@@ -705,6 +705,7 @@ Sorting is per order in the categories, not lexical order.
 Grouping by a categorical column also shows empty categories.
 
```
rather than the warning, can you pass the observed explicity

jreback commented on doc/source/user_guide/advanced.rst:
```diff
@@ -808,6 +808,7 @@ order is ``cab``).
 Groupby operations on the index will preserve the index nature as well.
 
 .. ipython:: python
+   :okwarning:
```
same

jreback commented on doc/source/user_guide/categorical.rst:
```diff
@@ -622,9 +622,10 @@ even if some categories are not present in the data:
+``DataFrame`` methods like :meth:`DataFrame.sum` also show "unused" categories:
 
 .. ipython:: python
+   :okwarning:
```
same, generally we want to fix any docs (except older what'snew where putting okwarning is ok)

jseabold replied:
Sure, but to be clear on this one, fixing this requires hard-coding observed=False (or True) [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/generic.py#L10548), which someone will eventually not love. Likely going to need to add another keyword, especially, when you change the default behavior down the line.

jreback replied:
if it does that's a problem
we should be able to hard code in the docs not the code

if there is some lower level where the warning is coming from then that is. problem and needs to be addressed

this is why we need to NOT just suppress warnings in tests as they could be hising a bug

jseabold replied:
There was one other place I had to do this. Any code path that uses a groupby internally is going to raise this warning with a categorical. Hopefully the test coverage is high enough not to make things super noisy for a lot of folks...

jseabold replied:
> if it does that's a problem

Every place you use groupby internally but don't allow passing all the keywords down that control it's type-dependent behavior, this can come up. If the code path is covered in tests then the warnings-as-errors should catch it. If not, then you may hear from some folks.

jreback commented on pandas/tests/groupby/aggregate/test_aggregate.py:
```diff
@@ -1060,6 +1060,7 @@ def test_groupby_get_by_index():
         ({"nr": "min"}, {"nr": [1, 5]}),
     ],
 )
+@pytest.mark.filterwarnings("ignore:Using 'observed:FutureWarning")
```
also ok with being explict about observed

jreback commented on pandas/tests/groupby/test_grouping.py:
```diff
@@ -702,6 +703,28 @@ def test_groupby_multiindex_level_empty(self):
         tm.assert_frame_equal(result, expected)
 
+    def test_default_observed_deprecated(self):
+        df = DataFrame([["A", 1, 1], ["A", 2, 1], ["B", 1, 1]], columns=["x", "y", "z"])
```
can you add this PR number here as a comment

jreback commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2058,6 +2067,12 @@ def agg(arr):
             foo.pivot_table("notpresent", "X", "Y", aggfunc=agg)
 
+    def test_pivot_table_observed_deprecated_default(self):
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
add a comment here with this PR number

jreback reviewed (changes requested):
generally ok with this, can you update the tests so we filter less (if its too crazy we can do as a followup).

jreback commented:
cc @jbrockmendel @jorisvandenbossche @TomAugspurger if any objections.

jreback commented:
> Finished the warnings whack-a-mole. Silencing warnings in years old what's new docs seemed like not the right thing to do, but it's in there. I don't know if you re-publish these and/or accept ipython directives in them, but if so they're likely not historically correct anymore.

yeah for old docs silencing is ok (or replacing with a code-block but); for the docs themselves we like to fix it (e.g. being explicit).

jseabold commented:
> can you update the tests so we filter less

The other alternative where it's not possible to hard-code observed is to add this everywhere, which is what I did at first. Seemed like the filtering pattern was used more elsewhere.

```
with tm.assert_produces_warning(FutureWarning if observed is None else None, ...)
    ...
```

jseabold committed:
Filter fewer warnings

jseabold committed:
Hard code observed behavior to silence warning. See #35967

jseabold committed:
PR in comment

jseabold committed:
Hardcode vs. filtering warning

jseabold committed:
Blacken

jreback commented:
internal use of groupby need to hardcode True (and maybe reindex if needed)

jseabold commented:
> internal use of groupby need to hardcode True (and maybe reindex if needed)

This would be a breaking change. The default now is False. Is that what you mean? Then every internal use of groupby will need a separate deprecation warning unless we leave them all as None (as they were) and catch the warnings. That's why I caught the warnings in the places where the user can't actually do anything but a future change will change this behavior. Really, everything that permits this behavior needs a new keyword but I'm not sure I have that in me tbh.

jreback commented:
i am not sure understand

we cannot have any internal uses showing warnings

you have to set these (can be False i guess as that is the current)

we cannot have warnings except by a user call

jseabold commented:
> i am not sure understand

[This](https://github.com/pandas-dev/pandas/blob/master/pandas/core/generic.py#L10548) code path is hit by a call to df.sum(level=1) and _if_ the dataframe has a categorical then it raised a warning. I hard-coded an `observed=False` to silence the warning. _But_ the user has no way to control this behavior by calling `sum`, because the `observed` keyword isn't exposed there. The user also has no way of knowing that this behavior will change whenever the default changes to `observed=True` for groupby. The only reason I knew this is because it raised a warning in the test suite/doc tests. If there are other untested code paths, then they may raise a warning. I didn't `grep` for groupby's to see where all this is possible. I also think adding a type-dependent `observed` keyword to every method that uses `groupby` internally smells, but I don't have a sensible alternative (other than Tom's new type).

jseabold committed:
Silence deprecation warning.

jseabold committed:
Hard code default behavior. See #35967

jseabold committed:
Hardcode default categorical behavior. See #35967

jseabold commented:
Feedback has been addressed. AFAICT there were only three places left with a `groupby` used internally without `observed` being exposed to users that the tests don't catch - `pd.merge_ordered`,  `DataFrame.value_counts`, and `DataFrame.boxplot`. I added the current hard-coded default to the first and last, and I added `observed=True` to the middle since the docs say it is by definition only the observed rows.

jorisvandenbossche commented on pandas/core/groupby/grouper.py:
```diff
@@ -31,6 +32,17 @@
+_observed_msg = textwrap.dedent(
+    """\
+Using 'observed=False', because grouping on a categorical. A future version
+of pandas will change to 'observed=True'.
```
This message here is not what I understood from the discussion in the PR? (https://github.com/pandas-dev/pandas/pull/35967#issuecomment-693685382)

TomAugspurger replied:
Yeah, I thought the suggestion was that this would *raise* in the future. So this warning message would be something like

```
Grouping by a categorical, but 'observed' was not specified.
Using 'observed=False', but in the future this will will raise
a ValueError. Pass 'observed=True' or 'observed=False' to
silence this warning.
```

Maybe with a link to the docs on what observed=True/False means.

jseabold replied:
Updated

jorisvandenbossche reviewed (changes requested):
@jseabold thanks a lot for working on this!

Now, I was reading the discussion in this PR, but now I don't fully understand what the PR is doing (well, it's clearly saying the `observed` default is deprecated and will become `True` as default in the future, but that's not my reading of the discussion?)

Will try to take a more detailed look at the actual changes in the coming days. 
But I don't think we need to rush this into 1.2 (unless it still takes a while before the RC is cut). Ideally we get also some more comments from people that were involved earlier in the discussion.

jseabold commented:
> I don't fully understand what the PR is doing

I see. The main thing this PR does is encourage users to be explicit about what they want to do with groupby in the presence of categoricals and starts us down the road of not possibly doing cartesian products by default. It doesn't change the behavior but warns that the default will change.

There were previous discussions on wanting to change the default eventually. We could end up in other situations though.

1. This could be the end state. If you have a categorical, you have to explicitly select True/False for `observed` or we yell at you w/ default False (refuse the temptation to guess).
2. Change the default. Yell at you until the default is changed.
3. Change the default, but continue to yell if you don't explicitly set it (refuse the temptation to guess and don't produce cartesian products sometimes).
4. Do nothing and wait for new types that determine their own behavior.

jreback commented on doc/source/user_guide/groupby.rst:
```diff
@@ -1264,6 +1264,7 @@ Categorical variables represented as instance of pandas's ``Categorical`` class
 can be used as group keys. If so, the order of the levels will be preserved:
 
 .. ipython:: python
+   :okwarning:
```
these are the ones that need to go. the problem is that this shows for *any* groupby that happens to hit this path, so i think users are going to have to be explicit on the keyword

jseabold replied:
Yeah, I just missed this one.

jseabold replied:
Fixed

jseabold committed:
Will raise in the future

jseabold commented:
I must have some global black config that conflicts with what y'all check with. Every time I save a file, it blackens it and I run into a conflict I need to fix.

jorisvandenbossche commented:
Regarding the future behaviour we want: always raising whenever you have a categorical in your groupby keys also might not be such a great user experience ..

Thinking about some other possible alternatives:

- Keep the default of `observed=False` for a single grouper, but deprecate + eventually change to `observed=True` for multiple groupers
  - This is basically the behaviour as it was before the `observed` keyword was introduced. And the memory issues (cartesian product combinatory explosion) only are an issue with multiple groupers, I think?
  - On the other hand, having different default behaviour for this depending on a single vs multiple grouper might also be confusing / surprising, and gives inconsistent behaviour ..
- Detect if there are unobserved categories, and only raise an error if in that case

jseabold commented:
Ha, yeah... I don't disagree about the UX. Everything about this smells - type dependent keywords, ambiguous desired behavior... Raising seems to be the only thing that refuses the temptation to guess though. In terms of lesser evils, it's good to have a guiding principle. 

Maybe like `ascending=False`, `observed=True` will just have to be a thing I type many times / day.

I kind of figured Tom would have new types done by the time that this comes down to actually raising an error, though I haven't really thought through how all of that will go.

jseabold commented:
This needs a decision. I'm going to put off merge conflicts chores until then.

jseabold commented:
Just blew my computer up again. Unstaling this PR.

mroeschke commented:
Thanks for the PR @jseabold but it appears that this discussion and PR has sufficiently stalled. It appears that this issue should be addressed in a way but may be better for discussion to resume on the path forward in https://github.com/pandas-dev/pandas/issues/30552 first. Closing but we can reopen this PR if this is the path the core devs decide on
